File: iosApp/iosApp/note_list/NoteListViewModel.swift
import Foundation
import shared

extension NoteListScreenView {
    
    @MainActor class NoteListViewModel: ObservableObject {
        
        private var noteDataSource: INoteDataSource? = nil
        private let searchNote = SearchNotes()
        
        private var notes = [Note]()
        @Published private(set) var filteredNote = [Note]()
        @Published private(set) var isSearchActive = false
        @Published var searchText = "" {
            didSet {
                self.filteredNote = searchNote.searchNotesExecute(notes: self.notes, query: searchText)
            }
        }
        
        init(noteDataSource: INoteDataSource? = nil) {
            self.noteDataSource = noteDataSource
        }
        
        func setNoteDataSource(noteDataSource: INoteDataSource) {
            self.noteDataSource = noteDataSource
        }
        
        func loadNotes() {
            noteDataSource?.getAllNotes(completionHandler: { notes, error in
                self.notes = notes ?? []
                self.filteredNote = self.notes
            })
        }
        
        func deleteById(id: Int64?) {
            if id != nil {
                noteDataSource?.deleteNoteById(id: id!, completionHandler: { error in
                    self.loadNotes()
                })
            }
        }
        
        func toggleIsSearchActive() {
            self.isSearchActive = !self.isSearchActive
            if !self.isSearchActive {
                searchText = ""
            }
        }
    }
}

File: androidApp/src/main/java/com/example/noteappkmm/android/vm/NoteListViewModel.kt
package com.example.noteappkmm.android.vm

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.noteappkmm.android.note_list.NoteListState
import com.example.noteappkmm.data.NoteDataSourceImpl
import com.example.noteappkmm.domain.note.INoteDataSource
import com.example.noteappkmm.domain.note.Note
import com.example.noteappkmm.domain.note.SearchNotes
import com.example.noteappkmm.domain.time.DateTimeUtil
import com.example.noteappkmm.presentation.RedOrangeHex
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class NoteListViewModel @Inject constructor(
    private val noteDataSource: INoteDataSource, private val savedState: SavedStateHandle
) : ViewModel() {
    private val searchNotes = SearchNotes()

    private val notes = savedState.getStateFlow("notes", emptyList<Note>())
    private val searchText = savedState.getStateFlow("searchText", "")
    private val isSearchActive = savedState.getStateFlow("isSearchActive", false)

    val state = combine(notes, searchText, isSearchActive) { notes, searchText, isSearchText ->
        NoteListState(
            notes = searchNotes.searchNotesExecute(notes, searchText),
            searchText = searchText,
            isSearchActive = isSearchText
        )
    }.stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), NoteListState())

    fun loadNotes() {
        viewModelScope.launch {
            savedState["notes"] = noteDataSource.getAllNotes()
        }
    }

    fun onSearchTextChange(text: String) {
        savedState["searchText"] = text
    }

    fun onSelectedToggleSearch() {
        savedState["isSearchActive"] = !isSearchActive.value
        if (!isSearchActive.value) {
            savedState["searchText"] = ""
        }
    }

    fun deleteNoteById(id: Long) {
        viewModelScope.launch {
            noteDataSource.deleteNoteById(id)
            loadNotes()
        }
    }
}